massa-event-cache/src/event_cache.rs

```diff
@@ -0,0 +1,490 @@
+use std::cmp::max;
+// std
+use nom::AsBytes;
+use std::path::Path;
+// third-party
+use crate::ser_deser::{
+    SCOutputEventDeserializer, SCOutputEventDeserializerArgs, SCOutputEventSerializer,
+};
+use massa_models::execution::EventFilter;
+use massa_models::output_event::SCOutputEvent;
+use massa_serialization::{DeserializeError, Deserializer, Serializer};
+use rocksdb::{DBIterator, IteratorMode, WriteBatch, DB};
+use tracing::debug;
+
+const OPEN_ERROR: &str = "critical: rocksdb open operation failed";
+const CRUD_ERROR: &str = "critical: rocksdb crud operation failed";
+const EVENT_DESER_ERROR: &str = "critical: event deserialization failed";
+
+/*
+/// Module key formatting macro
+#[macro_export]
+macro_rules! event_key {
+    ($event_slot:expr) => {
+        [&$event_slot.to_bytes()[..], &[MODULE_IDENT]].concat()
+    };
+}
+*/
+
+pub(crate) struct EventCache {
+    /// RocksDB database
+    db: DB,
+    /// How many entries are in the db. Count is initialized at creation time by iterating
+    /// over all the entries in the db then it is maintained in memory
+    entry_count: usize,
+    /// Maximum number of entries we want to keep in the db.
+    /// When this maximum is reached `snip_amount` entries are removed
+    max_entry_count: usize,
+    /// How many entries are removed when `entry_count` reaches `max_entry_count`
+    snip_amount: usize,
+    /// Event serializer
+    event_ser: SCOutputEventSerializer,
+    /// Event deserializer
+    event_deser: SCOutputEventDeserializer,
+}
+
+impl EventCache {
+    /// Create a new EventCache
+    pub fn new(
+        path: &Path,
+        max_entry_count: usize,
+        snip_amount: usize,
+        thread_count: u8,
+        max_recursive_call_depth: u16,
+        max_event_data_length: u64,
+    ) -> Self {
+        let db = DB::open_default(path).expect(OPEN_ERROR);
+
+        let mut event_cache = Self {
+            db,
+            entry_count: 0,
+            max_entry_count,
+            snip_amount,
+            event_ser: SCOutputEventSerializer::new(),
+            event_deser: SCOutputEventDeserializer::new(SCOutputEventDeserializerArgs {
+                thread_count,
+                max_recursive_call_depth,
+                max_event_data_length,
+            }),
+        };
+
+        event_cache.clear();
+        event_cache
+    }
+
+    fn clear(&mut self) {
+        let iter = self.db.iterator(IteratorMode::Start);
```

Comment:
There is a much faster way to clear than iterating over all elements. Check how we do it for the module cache or the state

Reply:
I've used Db::destroy to avoid the iteration